shared/js/passcode_helper.js

```diff
@@ -0,0 +1,161 @@
+/* exported PasscodeHelper */
+/* globals crypto, TextEncoder */
+(function(exports) {
+  'use strict';
+
+  const SET_DIGEST_VALUE = 'lockscreen.passcode-lock.digest.value';
+  const SET_DIGEST_SALT = 'lockscreen.passcode-lock.digest.salt';
+  const SET_DIGEST_ITERATIONS = 'lockscreen.passcode-lock.digest.iterations';
+  const SET_DIGEST_ALGORITHM = 'lockscreen.passcode-lock.digest.algorithm';
+  const DEFAULT_ALGORITHM = 'SHA-1'; //XXX Update to SHA-256 after bug 554827?
+  const DEFAULT_ITERATIONS = 1000;
+
+
+  var PasscodeHelper = function() {
+    /* Usage:
+     *      var ph = new PasscodeHelper();
+     *      ph.setPassccode(string) -> Promise => digest (or false)
+     *      ph.checkPasscode(string) -> Promise resolves to a boolean
+     * */
+
+
+    var _salt;
+    var _iterations;
+    var _algorithm;
+    var _testPass;
+
+    this._encode = function(str) {
+      return new TextEncoder('utf-8').encode(str);
+    };
+    this._toTypedArray = function(obj) {
+      // SettingsAPI doesnt like arrays and gives us an Object { 1: .., 2: .. }
+      var a = [];
+      for (var key in obj) {
+        a.push(obj[key]);
+      }
+      return new Uint8Array(a);
+    };
+    this._deriveBits = function(pwKey, salt, iterations, algorithm) {
+      var length = 256;
+      var params = {
+        name: 'PBKDF2',
+        hash: algorithm,
+        salt: salt,
+        iterations: iterations
+      };
+      return crypto.subtle.deriveBits(params, pwKey, length);
+    };
+
+    this._makeDigest = function(pass, salt, iterations, algorithm) {
+      var bytes = this._encode(pass);
+      return crypto.subtle.importKey('raw', bytes, 'PBKDF2', false, [
+        'deriveBits'
+      ]).then((pwKey) => {
+        return this._deriveBits(pwKey, salt, iterations, algorithm);
+      }).catch((error) => {
+        console.error('PasscodeHelper: _derive_bits() failed!', error);
+        throw error;
+      });
+    };
+
+    this.setPasscode = function(newPass) {
+      var lock = navigator.mozSettings.createLock();
+      var getFromSettings = Promise.all([
+        lock.get(SET_DIGEST_ITERATIONS),
+        lock.get(SET_DIGEST_ALGORITHM)
+      ]);
+      var digest = getFromSettings.then((values) => {
+        // Always generate a new salt.
+        _salt = crypto.getRandomValues(new Uint8Array(8));
+        /* the combined lock.get() makes it quite ugly. we get
+         an Array of objects, each with just one key/value,
+         which is the requested setting. let's destruct as follows:
+         */
+        var [digestIteration, digestAlgorithm] = values;
+        _iterations = parseInt(digestIteration[SET_DIGEST_ITERATIONS]) ||
+          DEFAULT_ITERATIONS;
+        _algorithm = typeof(digestAlgorithm[SET_DIGEST_ALGORITHM]) == 'string' ?
```

Comment:
nit: never use '=='. Use '===' instead. And, I would write this as:

`('string' === typeof digestAlgorithm[SET_DIGEST_ALGORITHM]) ? ... : ...`